File: app/src/main/java/com/kyilmaz/neurocomet/WallpaperEngine.kt
package com.kyilmaz.neurocomet

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

// =============================================================================
// NEURODIVERGENT-FRIENDLY WALLPAPER ENGINE
// Sensory-conscious animated backgrounds for Messages conversations.
// All rendering uses DrawScope (GPU-accelerated Canvas), zero recomposition.
// =============================================================================

/**
 * Available conversation wallpapers, each designed with sensory sensitivity in mind.
 */
enum class ConversationWallpaper(
    val displayName: String,
    val description: String,
    val emoji: String
) {
    NONE("None", "Plain background", "🚫"),
    CALM_OCEAN("Calm Ocean", "Slow-moving blue gradient waves", "🌊"),
    AURORA_BOREALIS("Aurora Borealis", "Gentle shifting northern lights", "🌌"),
    BREATHING_BUBBLES("Breathing Bubbles", "Soft pulsing circles for grounding", "🫧"),
    SOFT_RAIN("Soft Rain", "Gentle dots drifting downward", "🌧️"),
    DEEP_FOCUS("Deep Focus", "Minimal dark gradient, reduced stimulation", "🎯"),
    WARM_SUNSET("Warm Sunset", "Slow warm gradient cycle", "🌅"),
    STARFIELD("Starfield", "Twinkling stars on a calm night sky", "✨");

    companion object {
        fun fromKey(key: String?): ConversationWallpaper =
            entries.find { it.name == key } ?: NONE
    }
}

// =============================================================================
// Composable Modifier — attach to the Box behind the message list
// =============================================================================

/**
 * Applies a neurodivergent-friendly animated wallpaper behind the content.
 *
 * @param wallpaper The selected wallpaper preset.
 * @param isDark Whether the app is in dark mode.
 * @param reducedMotion If true, all animations are frozen at their initial frame.
 * @param dataSaver If true, treated as implicit reduced-motion (saves battery).
 */
@Composable
fun Modifier.conversationWallpaper(
    wallpaper: ConversationWallpaper,
    isDark: Boolean = isSystemInDarkTheme(),
    reducedMotion: Boolean = false,
    dataSaver: Boolean = false
): Modifier {
    if (wallpaper == ConversationWallpaper.NONE) return this

    val freeze = reducedMotion || dataSaver

    // ── Animation drivers (all use rememberInfiniteTransition for efficiency) ──
    val transition = rememberInfiniteTransition(label = "wallpaper_${wallpaper.name}")

    val slowPhase by transition.animateFloat(
        initialValue = 0f,
        targetValue = if (freeze) 0f else 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 8000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "slowPhase"
    )

    val breathPhase by transition.animateFloat(
        initialValue = 0f,
        targetValue = if (freeze) 0f else 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 4000, easing = LinearEasing),
            repeatMode = RepeatMode.Reverse
        ),
        label = "breathPhase"
    )

    val gentlePhase by transition.animateFloat(
        initialValue = 0f,
        targetValue = if (freeze) 0f else 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = 12000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "gentlePhase"
    )

    return this.drawBehind {
        when (wallpaper) {
            ConversationWallpaper.NONE -> { /* no-op */ }
            ConversationWallpaper.CALM_OCEAN -> drawCalmOcean(slowPhase, isDark)
            ConversationWallpaper.AURORA_BOREALIS -> drawAuroraBorealis(gentlePhase, isDark)
            ConversationWallpaper.BREATHING_BUBBLES -> drawBreathingBubbles(breathPhase, isDark)
            ConversationWallpaper.SOFT_RAIN -> drawSoftRain(slowPhase, isDark)
            ConversationWallpaper.DEEP_FOCUS -> drawDeepFocus(isDark)
            ConversationWallpaper.WARM_SUNSET -> drawWarmSunset(gentlePhase, isDark)
            ConversationWallpaper.STARFIELD -> drawStarfield(breathPhase, isDark)
        }
    }
}

// =============================================================================
// Individual Wallpaper Renderers (DrawScope extensions)
// =============================================================================

/**
 * Calm Ocean — Two overlapping sine-wave fills with translucent blues.
 */
private fun DrawScope.drawCalmOcean(phase: Float, isDark: Boolean) {
    val w = size.width
    val h = size.height

    // Base gradient
    val baseColors = if (isDark) {
        listOf(Color(0xFF0D1B2A), Color(0xFF1B2838), Color(0xFF0D253A))
    } else {
        listOf(Color(0xFFE3F2FD), Color(0xFFBBDEFB), Color(0xFFE1F5FE))
    }
    drawRect(brush = Brush.verticalGradient(baseColors))

    // Wave 1 — slow
    val wave1Alpha = if (isDark) 0.15f else 0.12f
    val wave1Color = if (isDark) Color(0xFF42A5F5) else Color(0xFF1976D2)
    val path1 = Path().apply {
        moveTo(0f, h * 0.65f)
        val phaseOffset = phase * 2f * PI.toFloat()
        for (x in 0..w.toInt() step 4) {
            val xf = x.toFloat()
            val y = h * 0.65f + sin(xf / w * 2f * PI.toFloat() + phaseOffset) * h * 0.04f
            lineTo(xf, y)
        }
        lineTo(w, h)
        lineTo(0f, h)
        close()
    }
    drawPath(path1, color = wave1Color.copy(alpha = wave1Alpha))

    // Wave 2 — offset
    val wave2Color = if (isDark) Color(0xFF64B5F6) else Color(0xFF42A5F5)
    val path2 = Path().apply {
        moveTo(0f, h * 0.72f)
        val phaseOffset2 = phase * 2f * PI.toFloat() + PI.toFloat() * 0.7f
        for (x in 0..w.toInt() step 4) {
            val xf = x.toFloat()
            val y = h * 0.72f + sin(xf / w * 3f * PI.toFloat() + phaseOffset2) * h * 0.03f
            lineTo(xf, y)
        }
        lineTo(w, h)
        lineTo(0f, h)
        close()
    }
    drawPath(path2, color = wave2Color.copy(alpha = wave1Alpha * 0.7f))
}

/**
 * Aurora Borealis — Vertical gradient with shifting color stops.
 */
private fun DrawScope.drawAuroraBorealis(phase: Float, isDark: Boolean) {
    val phaseRad = phase * 2f * PI.toFloat()

    // Background
    drawRect(color = if (isDark) Color(0xFF0A0E1A) else Color(0xFFF5F5F5))

    // Aurora bands — 3 overlapping vertical gradients with shifting positions
    val band1Y = size.height * (0.15f + sin(phaseRad) * 0.08f)
    val band1H = size.height * 0.25f
    val color1 = if (isDark) Color(0xFF00E676) else Color(0xFF81C784)
    drawRect(
        brush = Brush.verticalGradient(
            colors = listOf(Color.Transparent, color1.copy(alpha = 0.12f), Color.Transparent),
            startY = band1Y,
            endY = band1Y + band1H
        )
    )

    val band2Y = size.height * (0.30f + cos(phaseRad * 0.8f) * 0.06f)
    val band2H = size.height * 0.20f
    val color2 = if (isDark) Color(0xFF7C4DFF) else Color(0xFFBA68C8)
    drawRect(
        brush = Brush.verticalGradient(
            colors = listOf(Color.Transparent, color2.copy(alpha = 0.10f), Color.Transparent),
            startY = band2Y,
            endY = band2Y + band2H
        )
    )

    val band3Y = size.height * (0.50f + sin(phaseRad * 1.3f) * 0.05f)
    val band3H = size.height * 0.18f
    val color3 = if (isDark) Color(0xFF00BCD4) else Color(0xFF80CBC4)
    drawRect(
        brush = Brush.verticalGradient(
            colors = listOf(Color.Transparent, color3.copy(alpha = 0.08f), Color.Transparent),
            startY = band3Y,
            endY = band3Y + band3H
        )
    )
}

/**
 * Breathing Bubbles — Soft pulsing circles for anxiety grounding.
 * Follows a 4-second inhale/exhale rhythm.
 */
private fun DrawScope.drawBreathingBubbles(breathPhase: Float, isDark: Boolean) {
    // Background
    drawRect(color = if (isDark) Color(0xFF1A1A2E) else Color(0xFFF8F0FF))

    // breathPhase goes 0→1→0 (Reverse mode) matching inhale/exhale
    val scale = 0.7f + breathPhase * 0.3f
    val alpha = 0.06f + breathPhase * 0.08f

    data class Bubble(val cx: Float, val cy: Float, val baseRadius: Float, val color: Color)

    val bubbles = listOf(
        Bubble(0.25f, 0.30f, 0.12f, if (isDark) Color(0xFF7C4DFF) else Color(0xFFCE93D8)),
        Bubble(0.70f, 0.20f, 0.09f, if (isDark) Color(0xFF536DFE) else Color(0xFF90CAF9)),
        Bubble(0.50f, 0.55f, 0.14f, if (isDark) Color(0xFF00BFA5) else Color(0xFFA5D6A7)),
        Bubble(0.15f, 0.75f, 0.08f, if (isDark) Color(0xFFFF80AB) else Color(0xFFF8BBD0)),
        Bubble(0.80f, 0.65f, 0.10f, if (isDark) Color(0xFF64B5F6) else Color(0xFFBBDEFB))
    )

    bubbles.forEach { b ->
        val r = b.baseRadius * size.width * scale
        drawCircle(
            color = b.color.copy(alpha = alpha),
            radius = r,
            center = Offset(b.cx * size.width, b.cy * size.height)
        )
        // Inner glow
        drawCircle(
            color = b.color.copy(alpha = alpha * 0.5f),
            radius = r * 0.6f,
            center = Offset(b.cx * size.width, b.cy * size.height)
        )
    }
}

/**
 * Soft Rain — Gentle dots drifting downward, wrapping around.
 */
private fun DrawScope.drawSoftRain(phase: Float, isDark: Boolean) {
    // Background
    drawRect(color = if (isDark) Color(0xFF1A1D23) else Color(0xFFF0F4F8))

    val dropColor = if (isDark) Color(0xFF90CAF9) else Color(0xFF64B5F6)

    // Deterministic raindrop positions using seed-based math (no Random in draw)
    for (i in 0 until 24) {
        val seedX = (i * 0.618033988f) % 1f  // Golden ratio spacing
        val seedSpeed = 0.6f + (i % 5) * 0.1f
        val x = seedX * size.width
        val baseY = ((i * 0.3141592f + phase * seedSpeed) % 1f) * (size.height + 40f) - 20f
        val radius = 1.5f + (i % 3) * 0.8f
        val alpha = if (isDark) 0.20f else 0.15f

        drawCircle(
            color = dropColor.copy(alpha = alpha - (i % 4) * 0.03f),
            radius = radius,
            center = Offset(x, baseY)
        )
    }
}

/**
 * Deep Focus — Minimal dark gradient with no animation. Reduced stimulation.
 */
private fun DrawScope.drawDeepFocus(isDark: Boolean) {
    val colors = if (isDark) {
        listOf(Color(0xFF0D0D12), Color(0xFF1A1A24), Color(0xFF0D0D12))
    } else {
        listOf(Color(0xFFF5F5F5), Color(0xFFEEEEEE), Color(0xFFE8E8E8))
    }
    drawRect(
        brush = Brush.radialGradient(
            colors = colors,
            center = Offset(size.width * 0.5f, size.height * 0.4f),
            radius = size.width * 0.8f
        )
    )
}

/**
 * Warm Sunset — Slow warm gradient cycle with oranges, peaches, pinks.
 */
private fun DrawScope.drawWarmSunset(phase: Float, isDark: Boolean) {
    val phaseRad = phase * 2f * PI.toFloat()

    // Shift gradient center slowly
    val centerY = size.height * (0.35f + sin(phaseRad) * 0.1f)

    val colors = if (isDark) {
        val blend = (sin(phaseRad) + 1f) / 2f
        listOf(
            lerpColor(Color(0xFF1A0A0A), Color(0xFF2A1010), blend),
            lerpColor(Color(0xFF4A1A00), Color(0xFF3A1500), blend),
            lerpColor(Color(0xFF2A0D1A), Color(0xFF1A0A15), blend)
        )
    } else {
        val blend = (sin(phaseRad) + 1f) / 2f
        listOf(
            lerpColor(Color(0xFFFFF3E0), Color(0xFFFFECB3), blend),
            lerpColor(Color(0xFFFFCCBC), Color(0xFFF8BBD0), blend),
            lerpColor(Color(0xFFFFE0B2), Color(0xFFFFF9C4), blend)
        )
    }

    drawRect(
        brush = Brush.radialGradient(
            colors = colors,
            center = Offset(size.width * 0.5f, centerY),
            radius = size.width * 1.2f
        )
    )
}

/**
 * Starfield — Twinkling stars on a calm night sky.
 */
private fun DrawScope.drawStarfield(breathPhase: Float, isDark: Boolean) {
    // Background — deep night sky
    val bgColors = if (isDark) {
        listOf(Color(0xFF0A0A1A), Color(0xFF0F0F2A), Color(0xFF0A0A1A))
    } else {
        listOf(Color(0xFFE8EAF6), Color(0xFFE0E0F0), Color(0xFFEDE7F6))
    }
    drawRect(brush = Brush.verticalGradient(bgColors))

    val starColor = if (isDark) Color.White else Color(0xFF5C6BC0)

    // 30 stars with deterministic positions
    for (i in 0 until 30) {
        val x = ((i * 0.618033988f * 7f) % 1f) * size.width
        val y = ((i * 0.381966f * 13f) % 1f) * size.height
        val baseRadius = 0.8f + (i % 4) * 0.4f

        // Twinkle: different stars phase-shift differently
        val twinkle = sin(breathPhase * PI.toFloat() + i * 0.5f)
        val alpha = if (isDark) {
            0.3f + twinkle * 0.4f
        } else {
            0.15f + twinkle * 0.2f
        }

        drawCircle(
            color = starColor.copy(alpha = alpha.coerceIn(0.05f, 0.9f)),
            radius = baseRadius * (0.8f + twinkle * 0.2f),
            center = Offset(x, y)
        )
    }
}

// =============================================================================
// Utility
// =============================================================================

/** Simple linear interpolation between two colors. */
private fun lerpColor(a: Color, b: Color, t: Float): Color {
    return Color(
        red = a.red + (b.red - a.red) * t,
        green = a.green + (b.green - a.green) * t,
        blue = a.blue + (b.blue - a.blue) * t,
        alpha = a.alpha + (b.alpha - a.alpha) * t
    )
}

File: flutter_app/lib/widgets/chat/wallpaper_engine.dart
import 'dart:math' as math;
import 'package:flutter/material.dart';

enum ConversationWallpaper {
  none,
  calmOcean,
  auroraBorealis,
  breathingBubbles,
  softRain,
  deepFocus,
  warmSunset,
  starfield,
}

class WallpaperPainter extends CustomPainter {
  final ConversationWallpaper wallpaper;
  final double phase;
  final bool isDark;

  WallpaperPainter({
    required this.wallpaper,
    required this.phase,
    required this.isDark,
  });

  @override
  void paint(Canvas canvas, Size size) {
    switch (wallpaper) {
      case ConversationWallpaper.none:
        break;
      case ConversationWallpaper.calmOcean:
        _drawCalmOcean(canvas, size);
        break;
      case ConversationWallpaper.auroraBorealis:
        _drawAuroraBorealis(canvas, size);
        break;
      case ConversationWallpaper.breathingBubbles:
        _drawBreathingBubbles(canvas, size);
        break;
      case ConversationWallpaper.softRain:
        _drawSoftRain(canvas, size);
        break;
      case ConversationWallpaper.deepFocus:
        _drawDeepFocus(canvas, size);
        break;
      case ConversationWallpaper.warmSunset:
        _drawWarmSunset(canvas, size);
        break;
      case ConversationWallpaper.starfield:
        _drawStarfield(canvas, size);
        break;
    }
  }

  void _drawCalmOcean(Canvas canvas, Size size) {
    final w = size.width;
    final h = size.height;

    final baseColors = isDark
        ? [const Color(0xFF0D1B2A), const Color(0xFF1B2838), const Color(0xFF0D253A)]
        : [const Color(0xFFE3F2FD), const Color(0xFFBBDEFB), const Color(0xFFE1F5FE)];

    final paint = Paint()
      ..shader = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: baseColors,
      ).createShader(Rect.fromLTWH(0, 0, w, h));
    canvas.drawRect(Rect.fromLTWH(0, 0, w, h), paint);

    final wave1Alpha = isDark ? 0.15 : 0.12;
    final wave1Color = isDark ? const Color(0xFF42A5F5) : const Color(0xFF1976D2);
    final phaseOffset = phase * 2 * math.pi;

    final path1 = Path()..moveTo(0, h * 0.65);
    for (double x = 0; x <= w; x += 4) {
      final y = h * 0.65 + math.sin(x / w * 2 * math.pi + phaseOffset) * h * 0.04;
      path1.lineTo(x, y);
    }
    path1.lineTo(w, h);
    path1.lineTo(0, h);
    path1.close();
    canvas.drawPath(path1, Paint()..color = wave1Color.withValues(alpha: wave1Alpha));

    final wave2Color = isDark ? const Color(0xFF64B5F6) : const Color(0xFF42A5F5);
    final phaseOffset2 = phase * 2 * math.pi + math.pi * 0.7;
    final path2 = Path()..moveTo(0, h * 0.72);
    for (double x = 0; x <= w; x += 4) {
      final y = h * 0.72 + math.sin(x / w * 3 * math.pi + phaseOffset2) * h * 0.03;
      path2.lineTo(x, y);
    }
    path2.lineTo(w, h);
    path2.lineTo(0, h);
    path2.close();
    canvas.drawPath(path2, Paint()..color = wave2Color.withValues(alpha: wave1Alpha * 0.7));
  }

  void _drawAuroraBorealis(Canvas canvas, Size size) {
    final phaseRad = phase * 2 * math.pi;
    canvas.drawRect(Rect.fromLTWH(0, 0, size.width, size.height), Paint()..color = isDark ? const Color(0xFF0A0E1A) : const Color(0xFFF5F5F5));

    final band1Y = size.height * (0.15 + math.sin(phaseRad) * 0.08);
    final band1H = size.height * 0.25;
    final color1 = isDark ? const Color(0xFF00E676) : const Color(0xFF81C784);
    canvas.drawRect(
      Rect.fromLTWH(0, band1Y, size.width, band1H),
      Paint()..shader = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [Colors.transparent, color1.withValues(alpha: 0.12), Colors.transparent],
      ).createShader(Rect.fromLTWH(0, band1Y, size.width, band1H)),
    );

    final band2Y = size.height * (0.30 + math.cos(phaseRad * 0.8) * 0.06);
    final band2H = size.height * 0.20;
    final color2 = isDark ? const Color(0xFF7C4DFF) : const Color(0xFFBA68C8);
    canvas.drawRect(
      Rect.fromLTWH(0, band2Y, size.width, band2H),
      Paint()..shader = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [Colors.transparent, color2.withValues(alpha: 0.10), Colors.transparent],
      ).createShader(Rect.fromLTWH(0, band2Y, size.width, band2H)),
    );

    final band3Y = size.height * (0.50 + math.sin(phaseRad * 1.3) * 0.05);
    final band3H = size.height * 0.18;
    final color3 = isDark ? const Color(0xFF00BCD4) : const Color(0xFF80CBC4);
    canvas.drawRect(
      Rect.fromLTWH(0, band3Y, size.width, band3H),
      Paint()..shader = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [Colors.transparent, color3.withValues(alpha: 0.08), Colors.transparent],
      ).createShader(Rect.fromLTWH(0, band3Y, size.width, band3H)),
    );
  }

  void _drawBreathingBubbles(Canvas canvas, Size size) {
    canvas.drawRect(Rect.fromLTWH(0, 0, size.width, size.height), Paint()..color = isDark ? const Color(0xFF1A1A2E) : const Color(0xFFF8F0FF));

    final scale = 0.7 + phase * 0.3;
    final alpha = 0.06 + phase * 0.08;

    final bubbles = [
      (0.25, 0.30, 0.12, isDark ? const Color(0xFF7C4DFF) : const Color(0xFFCE93D8)),
      (0.70, 0.20, 0.09, isDark ? const Color(0xFF536DFE) : const Color(0xFF90CAF9)),
      (0.50, 0.55, 0.14, isDark ? const Color(0xFF00BFA5) : const Color(0xFFA5D6A7)),
      (0.15, 0.75, 0.08, isDark ? const Color(0xFFFF80AB) : const Color(0xFFF8BBD0)),
      (0.80, 0.65, 0.10, isDark ? const Color(0xFF64B5F6) : const Color(0xFFBBDEFB)),
    ];

    for (final b in bubbles) {
      final r = b.$3 * size.width * scale;
      final center = Offset(b.$1 * size.width, b.$2 * size.height);
      canvas.drawCircle(center, r, Paint()..color = b.$4.withValues(alpha: alpha));
      canvas.drawCircle(center, r * 0.6, Paint()..color = b.$4.withValues(alpha: alpha * 0.5));
    }
  }

  void _drawSoftRain(Canvas canvas, Size size) {
    canvas.drawRect(Rect.fromLTWH(0, 0, size.width, size.height), Paint()..color = isDark ? const Color(0xFF1A1D23) : const Color(0xFFF0F4F8));
    final dropColor = isDark ? const Color(0xFF90CAF9) : const Color(0xFF64B5F6);

    for (int i = 0; i < 24; i++) {
      final seedX = (i * 0.618033988) % 1.0;
      final seedSpeed = 0.6 + (i % 5) * 0.1;
      final x = seedX * size.width;
      final baseY = ((i * 0.3141592 + phase * seedSpeed) % 1.0) * (size.height + 40) - 20;
      final radius = 1.5 + (i % 3) * 0.8;
      final alpha = isDark ? 0.20 : 0.15;

      canvas.drawCircle(Offset(x, baseY), radius, Paint()..color = dropColor.withValues(alpha: alpha - (i % 4) * 0.03));
    }
  }

  void _drawDeepFocus(Canvas canvas, Size size) {
    final colors = isDark
        ? [const Color(0xFF0D0D12), const Color(0xFF1A1A24), const Color(0xFF0D0D12)]
        : [const Color(0xFFF5F5F5), const Color(0xFFEEEEEE), const Color(0xFFE8E8E8)];
    canvas.drawRect(
      Rect.fromLTWH(0, 0, size.width, size.height),
      Paint()..shader = RadialGradient(
        colors: colors,
        center: const Alignment(0, -0.2), // Adjusted from 0.5, 0.4
        radius: 0.8,
      ).createShader(Rect.fromLTWH(0, 0, size.width, size.height)),
    );
  }

  void _drawWarmSunset(Canvas canvas, Size size) {
    final phaseRad = phase * 2 * math.pi;
    final centerY = size.height * (0.35 + math.sin(phaseRad) * 0.1);
    final blend = (math.sin(phaseRad) + 1.0) / 2.0;

    final colors = isDark
        ? [
            Color.lerp(const Color(0xFF1A0A0A), const Color(0xFF2A1010), blend)!,
            Color.lerp(const Color(0xFF4A1A00), const Color(0xFF3A1500), blend)!,
            Color.lerp(const Color(0xFF2A0D1A), const Color(0xFF1A0A15), blend)!,
          ]
        : [
            Color.lerp(const Color(0xFFFFF3E0), const Color(0xFFFFECB3), blend)!,
            Color.lerp(const Color(0xFFFFCCBC), const Color(0xFFF8BBD0), blend)!,
            Color.lerp(const Color(0xFFFFE0B2), const Color(0xFFFFF9C4), blend)!,
          ];

    canvas.drawRect(
      Rect.fromLTWH(0, 0, size.width, size.height),
      Paint()..shader = RadialGradient(
        colors: colors,
        center: Alignment(0, (centerY / size.height) * 2 - 1),
        radius: 1.2,
      ).createShader(Rect.fromLTWH(0, 0, size.width, size.height)),
    );
  }

  void _drawStarfield(Canvas canvas, Size size) {
    final bgColors = isDark
        ? [const Color(0xFF0A0A1A), const Color(0xFF0F0F2A), const Color(0xFF0A0A1A)]
        : [const Color(0xFFE8EAF6), const Color(0xFFE0E0F0), const Color(0xFFEDE7F6)];

    canvas.drawRect(
      Rect.fromLTWH(0, 0, size.width, size.height),
      Paint()..shader = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: bgColors,
      ).createShader(Rect.fromLTWH(0, 0, size.width, size.height)),
    );

    final starColor = isDark ? Colors.white : const Color(0xFF5C6BC0);

    for (int i = 0; i < 30; i++) {
      final x = ((i * 0.618033988 * 7.0) % 1.0) * size.width;
      final y = ((i * 0.381966 * 13.0) % 1.0) * size.height;
      final baseRadius = 0.8 + (i % 4) * 0.4;

      final twinkle = math.sin(phase * math.pi + i * 0.5);
      final alpha = isDark ? (0.3 + twinkle * 0.4) : (0.15 + twinkle * 0.2);

      canvas.drawCircle(
        Offset(x, y),
        baseRadius * (0.8 + twinkle * 0.2),
        Paint()..color = starColor.withValues(alpha: alpha.clamp(0.05, 0.9)),
      );
    }
  }

  @override
  bool shouldRepaint(WallpaperPainter oldDelegate) =>
      oldDelegate.wallpaper != wallpaper ||
      oldDelegate.phase != phase ||
      oldDelegate.isDark != isDark;
}

class ConversationWallpaperWidget extends StatefulWidget {
  final ConversationWallpaper wallpaper;
  final bool reducedMotion;

  const ConversationWallpaperWidget({
    super.key,
    required this.wallpaper,
    this.reducedMotion = false,
  });

  @override
  State<ConversationWallpaperWidget> createState() => _ConversationWallpaperWidgetState();
}

class _ConversationWallpaperWidgetState extends State<ConversationWallpaperWidget>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(seconds: 8),
      vsync: this,
    );
    if (!widget.reducedMotion && widget.wallpaper != ConversationWallpaper.none) {
      _controller.repeat();
    }
  }

  @override
  void didUpdateWidget(ConversationWallpaperWidget oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.reducedMotion != oldWidget.reducedMotion || widget.wallpaper != oldWidget.wallpaper) {
      if (widget.reducedMotion || widget.wallpaper == ConversationWallpaper.none) {
        _controller.stop();
      } else {
        _controller.repeat();
      }
    }
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    if (widget.wallpaper == ConversationWallpaper.none) return const SizedBox.shrink();

    final isDark = Theme.of(context).brightness == Brightness.dark;

    return AnimatedBuilder(
      animation: _controller,
      builder: (context, child) {
        return CustomPaint(
          painter: WallpaperPainter(
            wallpaper: widget.wallpaper,
            phase: _controller.value,
            isDark: isDark,
          ),
          size: Size.infinite,
        );
      },
    );
  }
}
